1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py

```diff
@@ -0,0 +1,320 @@
+gold_crops = root_dir / "vision_nuclear_speckle_prediction/gold_cropped_nuclei"
+
+# Paths to original nuclear speckle data
+data_dir = root_dir / "nuclear_speckles_data"
+nuclear_mask_dir = (data_dir / "Nuclear_masks").resolve(strict=True)
+sc_profiles_path = list((data_dir / "Preprocessed_data/single_cell_profiles").resolve(strict=True).glob("*feature_selected*.parquet"))
+
+# Load single-cell profile data
+scdfs = [pd.read_parquet(sc_path) for sc_path in sc_profiles_path if sc_path.is_file()]
+scdfs = pd.concat(scdfs, axis=0).reset_index(drop=True)
+
+
+# # Outputs
+
+# In[ ]:
+
+
+figure_path = pathlib.Path("fnet_validation_images_temp")
+figure_path.mkdir(parents=True, exist_ok=True)
+
+metrics_path = pathlib.Path("metrics")
+metrics_path.mkdir(parents=True, exist_ok=True)
+
+model_path = pathlib.Path("model")
+model_path.mkdir(parents=True, exist_ok=True)
+
+
+# In[ ]:
+
+
+description = "Here we leverage the 2d fnet architecture in https://doi.org/10.1038/s41592-018-0111-2 to predict the GOLD stain from cropped DAPI Nuclei Images. We retain all pixel values in the cropped images"
+mlflow.set_tag("mlflow.note.content", description)
+
+
+# # Image Generation Functions
+
+# In[ ]:
+
+
+def format_img(_tensor_img):
+    """Reshapes an image and rescales pixel values from the StandardScaler transform."""
+
+    mean = trainer.val_dataset.dataset.input_transform[0].mean
+    std = trainer.val_dataset.dataset.input_transform[0].std
+
+    return (torch.squeeze(_tensor_img) * std + mean).to(torch.uint16).cpu().numpy()
+
+
+# In[ ]:
+
+
+def evaluate_and_format_imgs(_input, _target):
+
+    single_input = input.unsqueeze(1).to(device)
+    single_target = target.to(device)
+
+    model.eval()
+    with torch.no_grad():
+        # Forward Pass
+        output = model(single_input)
+
+    return format_img(single_input), format_img(single_target), format_img(output)
+
+
+# # Initialize and Train Model
+
+# In[ ]:
+
+
+transforms = A.Compose([
+    StandardScaler(_always_apply=True),
+    CropNPixels(_pixel_count=1, _always_apply=True)
+])
+
+img_dataset = ImageDataset(
+    _input_dir=treated_dapi_crops,
+    _target_dir=gold_crops,
+    _input_transform=transforms,
+    _target_transform=transforms
+)
+
+
+# In[ ]:
+
+
+model = Net()
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+
+model.to(device)
+
+
+# In[ ]:
+
+
+optim_params = {
+    "lr": 1e-3,
+    "betas": (0.5, 0.999)
+}
+
+optimizer = optim.Adam(
+    model.parameters(),
+    **optim_params
+)
+
+mlflow.log_param("Optimizer", "ADAM")
+mlflow.log_params(optim_params)
+
+
+# In[ ]:
+
+
+# These keys will be in the names of the logged losses
+tracked_losses = {
+    "mse loss": nn.MSELoss(),
+    "mae loss": nn.L1Loss()
+}
+
+backprop_loss_name = "mae loss"
+
+mlflow.log_param("Training Loss", backprop_loss_name)
+
+
+# In[ ]:
+
+
+trainer_params = {
+    "_batch_size": 32,
+    "_epochs": 2_000,
+    "_patience": 15
+}
```

Comment:
Could these params be defined in a single nested dict or in say a `toml` file?

Reply:
Probably, I was passing parameters using argparse, but this didn't work when running a jupyter notebook. I could see this being faster for training. On gh though, I think this is more of a preference, but maybe this will change later